File: Prediction of Diabetes.py
# Analysizing the business problem
# Target Variable = Class Variable - Predicting the Diabetes
# Independent Variables = Other Factors 

import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

# Importing the dataset
Diabetes = pd.read_csv("C:/Users/personal/Desktop/Diabetes.csv")
Diabetes.columns

# Renaming the column names
Diabetes.columns = "NP","PGC","BP","SFT","SI","BMI","DPF","Age","CV"

# Creation of a Dummy Variable for output
lb = LabelEncoder()
Diabetes["CV"] = lb.fit_transform(Diabetes["CV"])

# Normalization function 
def norm_func(i):
    x = (i-i.min())	/ (i.max()-i.min())
    return (x)

# Normalized data frame (considering the numerical part of data)
Diabetes_n = norm_func(Diabetes.iloc[:, 0:9])
Diabetes.info()
Diabetes_n.describe()

Diabetes_n
X = np.array(Diabetes_n.iloc[:,0:8]) # Predictors 
Y = np.array(Diabetes_n['CV']) # Target 

# Model Building
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3)

from sklearn.ensemble import RandomForestClassifier

help(RandomForestClassifier)
rf = RandomForestClassifier(n_jobs=2, n_estimators=15, criterion="entropy")

rf.fit(X_train, Y_train) # Fitting RandomForestClassifier model from sklearn.ensemble  

pred = rf.predict(X_test)

from sklearn.metrics import accuracy_score

pd.crosstab(Y_test, pred, rownames=['Actual'], colnames= ['Predictions'])

print(accuracy_score(Y_test, pred)) #0.77

# test accuracy
test_acc2 = np.mean(rf.predict(X_test)==Y_test)
test_acc2 # 0.77

# train accuracy 
train_acc2 = np.mean(rf.predict(X_train)==Y_train)
train_acc2 # 0.99

# Conclusion
# From the train and test accuracy we can say that still the model is overfit

File: Prediction of Diabetes.R
# Random Forest Analysis for Diabetes Data Set

str(Diabetes)
attach(Diabetes)

table(Class.variable)

round(prop.table(table(Class.variable))*100, 1)


#Create a function to normalize the data
normFunc <- function(x){(x-mean(x, na.rm = T))/sd(x, na.rm = T)}
Diabetes[1:8] <- apply(Diabetes[1:8], 2, normFunc)
View(Diabetes)

Diabetes_rand <- Diabetes[order(runif(768)), ]
View(Diabetes_rand)

# create training and test data
train <- Diabetes_rand[1:500, ]
test <- Diabetes_rand[501:768, ]


# Building a random forest model on training data 
install.packages("randomForest")
library(randomForest)

?randomForest
Diabetes_forest <- randomForest(train$Class.variable ~ ., data = train, importance = TRUE)
plot(Diabetes_forest)

# Test Data Accuracy
test_acc <- mean(test$Class.variable == predict(Diabetes_forest, newdata=test))
test_acc # 0.776

# Train Data Accuracy
train_acc <- mean(train$Class.variable == predict(Diabetes_forest, data=train))
train_acc # 0.76
              


